include/mrdox/Metadata/Template.hpp

```diff
@@ -14,39 +14,254 @@
 #define MRDOX_METADATA_TEMPLATE_HPP
 
 #include <mrdox/Platform.hpp>
-#include <mrdox/Metadata/TemplateArg.hpp>
-#include <mrdox/Metadata/TemplateParam.hpp>
+#include <mrdox/ADT/Optional.hpp>
+#include <mrdox/Metadata/Type.hpp>
+#include <optional>
+#include <string>
+#include <vector>
 
 namespace clang {
 namespace mrdox {
 
+enum class TParamKind : int
+{
+    None = 0, // for bitstream
+    // template type parameter, e.g. "typename T" or "class T"
+    Type,
+    // template non-type parameter, e.g. "int N" or "auto N"
+    NonType,
+    // template template parameter, e.g. "template<typename> typename T"
+    Template,
+};
+
+struct TParam;
+
+struct TypeTParam
+{
+    /** Default type for the type template parameter */
+    std::optional<TypeInfo> Default;
+};
+
+struct NonTypeTParam
+{
+    /** Type of the non-type template parameter */
+    TypeInfo Type;
+    // Non-type template parameter default value (if any)
+    Optional<std::string> Default;
+};
+
+struct TemplateTParam
+{
+    /** Template parameters for the template template parameter */
+    std::vector<TParam> Params;
+    /** Non-type template parameter default value (if any) */
+    Optional<std::string> Default;
+};
+
+// ----------------------------------------------------------------
+
+struct TParam
+{
+private:
+    union Variant
+    {
+        Variant() { }
+        ~Variant() { }
+
+        TypeTParam Type;
+        NonTypeTParam NonType;
+        TemplateTParam Template;
+    } Variant_;
+
+    void destroy() const noexcept;
+
+    template<typename T>
+    void
+    construct(T&& other)
+    {
+        switch(other.Kind)
+        {
+        case TParamKind::Type:
+            emplace<TypeTParam>(
+                std::forward<T>(other).Variant_.Type);
+            break;
+        case TParamKind::NonType:
+            emplace<NonTypeTParam>(
+                std::forward<T>(other).Variant_.NonType);
+            break;
+        case TParamKind::Template:
+            emplace<TemplateTParam>(
+                std::forward<T>(other).Variant_.Template);
+            break;
+        default:
+            break;
+        }
+    }
+
+public:
+    /** The kind of template parameter this is. */
+    TParamKind Kind = TParamKind::None;
```

Comment:
is TParamKind::None a valid value?

Reply:
Yes, it is used to indicate that the type of template parameter has not been determined. It ensures that we don't call the wrong destructor in case of an error when reading bitcode.